### DISCUSSION TITLE
Transaction count after Execute indicates a mismatching number of BEGIN and COMMIT statements

### DISCUSSION THREAD

**Discussion by BDChinmayi** [serenity-is/Serenity#6736] [Help]
While editing purchases, it should update the values in database when clicking the button Apply Changes.
But instead of that it displays error message as 'Transaction count after Execute indicates a mismatching number of BEGIN and COMMIT statements. Previous count = 1, current count = 2.'
![Image1](https://user-images.githubusercontent.com/126105942/226894272-8de727ea-0800-447f-89f3-b146a4d6b6dc.jpg)

### What did you expect to happen?

How to rectify this error, please guide me

### How to reproduce?

1. Add new purchase detail by clicking the link New Purchases, if there is no company name please add company name.
2. Display the purchases by clicking the the menu item Purchases
3. And click on id to edit the purchase detail, made some changes in the displayed dialog box and click the button Apply Changes button. 
4. It displays the error message and unable to update the purchase detail changes.

### What Serenity Nuget Versions are you seeing the problem on? (separated by comma)

6.3.0.131

### Relevant log output

```console
An exception of type 'Microsoft.Data.SqlClient.SqlException' occurred in Serenity.Net.Data.dll but was not handled in user code
Transaction count after EXECUTE indicates a mismatching number of BEGIN and COMMIT statements. Previous count = 1, current count = 2.
```

### Database type and version

SQL Server 2019

### On which operating system do you see the problem?

Windows 10

### On which browsers do you see the problem?

Microsoft Edge

### On what version of the browsers do you see the problem?

Version 111.0.1661.44

### Additional information

I am using ASP dot net core 6.0

**Comment by volkanceylan** [comment]
Check your triggers, you are probably starting a transaction inside the transaction started by the unit of work

**Reply by BDChinmayi** [reply]
I didn't get this, please explain.

**Reply by edwardch** [reply]
If you have a trigger on a table(s) you are updating, the transaction inside the trigger can interfere with the count of the SQL statements.

Find your triggers

```sql
SELECT 
     sysobjects.name AS trigger_name 
    ,USER_NAME(sysobjects.uid) AS trigger_owner 
    ,s.name AS table_schema 
    ,OBJECT_NAME(parent_obj) AS table_name 
    ,OBJECTPROPERTY( id, 'ExecIsUpdateTrigger') AS isupdate 
    ,OBJECTPROPERTY( id, 'ExecIsDeleteTrigger') AS isdelete 
    ,OBJECTPROPERTY( id, 'ExecIsInsertTrigger') AS isinsert 
    ,OBJECTPROPERTY( id, 'ExecIsAfterTrigger') AS isafter 
    ,OBJECTPROPERTY( id, 'ExecIsInsteadOfTrigger') AS isinsteadof 
    ,OBJECTPROPERTY(id, 'ExecIsTriggerDisabled') AS [disabled] 
FROM sysobjects 

INNER JOIN sysusers 
    ON sysobjects.uid = sysusers.uid 

INNER JOIN sys.tables t 
    ON sysobjects.parent_obj = t.object_id 

INNER JOIN sys.schemas s 
    ON t.schema_id = s.schema_id 

WHERE sysobjects.type = 'TR' 
```

Open/modify the trigger, and you can add `SET NOCOUNT ON;` at the top of the body of your trigger statement.

**Reply by BDChinmayi** [reply]
There is no trigger statement. All are handled inside the coding. I have a db query file which is used to create tables and insert the values into it. I unable to attach that file here.  So kindly guide me to rectify this error. And also query file has that statement SET NOCOUNT ON;

**Comment by BDChinmayi** [comment]
The error is rectified. As per your guidance I have searched the trigger content. In the trigger, BEGIN TRANSACTION is occurred twice but COMMIT is occurred once, I removed BEGIN TRANSACTION from the trigger, After it works fine. Thank you so much sir for your guidance.
